modules/navigation/godot_navigation_server.cpp

```diff
@@ -639,6 +639,98 @@ void GodotNavigationServer::process(real_t p_delta_time) {
 	}
 }
 
+/////////////////////////////////////
+#ifdef MODULE_NAVIGATION_ENABLED
+gd::PathQueryResult GodotNavigationServer::_query_path(const gd::PathQueryParameters &p_parameters) const {
+	gd::PathQueryResult r_query_result;
+
+	const NavMap *map = map_owner.get_or_null(p_parameters.map);
+	ERR_FAIL_COND_V(map == nullptr, r_query_result);
+
+	// run the pathfinding
+
+	if (p_parameters.pathfinding_algorithm == gd::PathfindingAlgorithm::PATHFINDING_ALGORITHM_ASTAR) {
+		r_query_result.path = map->get_path(p_parameters.start_position, p_parameters.target_position, p_parameters.optimize, p_parameters.navigation_layers);
+	} else {
+		return r_query_result;
+	}
+
+	// add path postprocessing
+
+	if (!r_query_result.path.is_empty() && p_parameters.simplify_path) {
+		_simplify_path(p_parameters, r_query_result);
+	}
+
+	// add calculated path stats
+
+	if (r_query_result.path.size() > 1 && p_parameters.calculate_path_length) {
+		_calculate_path_length(p_parameters, r_query_result);
+	}
+
+	return r_query_result;
+}
+
+void GodotNavigationServer::_simplify_path(const gd::PathQueryParameters &p_parameters, gd::PathQueryResult &p_query_result) const {
+	Vector<Vector3> simplified_path;
+
+	int path_index = 0;
+	Vector3 new_direction;
+	Vector3 past_direction;
+	bool has_past_direction = false;
+	float margin_limit = 1.0 - p_parameters.simplify_path_margin;
+	float dot_product = 0.0;
+	bool is_straight = false;
+
+	while (path_index < p_query_result.path.size()) {
+		const Vector3 &path_pos = p_query_result.path[path_index];
+
+		if (path_index + 1 < p_query_result.path.size()) {
+			const Vector3 &path_pos_next = p_query_result.path[path_index + 1];
+
+			if (path_pos.is_equal_approx(path_pos_next)) {
+				path_index += 1;
+				continue;
+			}
+
+			new_direction = (path_pos_next - path_pos).normalized();
+			is_straight = false;
+
+			if (has_past_direction) {
+				dot_product = past_direction.dot(new_direction);
+
+				if (dot_product > margin_limit) {
+					is_straight = true;
+				}
+			}
+			past_direction = new_direction;
+			has_past_direction = true;
+
+			if (!is_straight) {
+				simplified_path.push_back(path_pos);
+			}
+		} else {
+			simplified_path.push_back(path_pos);
+		}
+
+		path_index += 1;
+	}
+
+	p_query_result.path = simplified_path;
+}
+
```

Comment:
It would be nice to see some graphical examples of paths where this is a problem currently. I'm slightly concerned that using dot product in this way could cause agents to get stuck in some situations, while also not being super intuitive for users.

Am suspecting it may be better here to wait on the ray tracing functionality for the navmesh, because that will enable auto-simplification of paths without cutting corners etc, and won't need parameterization by the user. Or else use a very conservative hard coded dot product (or similar) here as a placeholder and an on switch for path simplification.

But unless this is super pressing, it may be better to wait on this.

Reply:
Will split and wait, the path simplifying is not of core importance for this pr and can be added later in fashion.